example/dynamic.go

```diff
@@ -0,0 +1,140 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/inconshreveable/log15"
+	"github.com/ngrok/libngrok-go"
+)
+
+func unwrap[T any](out T, err error) T {
+	if err != nil {
+		log15.Error("unwrapped error", "error", err)
+		os.Exit(1)
+	}
+
+	return out
+}
+
+func httpResp(w http.ResponseWriter, code int, msg string, args ...any) {
+	w.WriteHeader(code)
+	fmt.Fprintf(w, msg, args...)
+}
+
+func main() {
+	log15.Root().SetHandler(log15.LvlFilterHandler(log15.LvlInfo, log15.StdoutHandler))
+
+	restart := true
+	domain := ""
+
+	ctx := context.Background()
+
+	for restart {
+		sess := unwrap(libngrok.Connect(ctx, libngrok.ConnectOptions().
+			WithLog15(log15.Root()).
+			WithRemoteCallbacks(libngrok.RemoteCallbacks{
+				OnStop: func(ctx context.Context, sess libngrok.Session) error {
+					restart = false
+					log15.Info("exiting due to remote request")
+					return nil
+				},
+				OnRestart: func(ctx context.Context, sess libngrok.Session) error {
+					log15.Info("restarting due to remote request")
+					return nil
+				},
+			}).
+			WithAuthToken(os.Getenv("NGROK_TOKEN")),
+		))
+
+		tun := unwrap(sess.StartTunnel(ctx, libngrok.HTTPOptions().
+			WithDomain(domain).
+			WithForwardsTo("tunnel management"),
+		))
+
+		u := unwrap(url.Parse(tun.URL()))
+		domain = u.Hostname()
+
+		log15.Info("managing tunnels", "url", tun.URL())
+
+		log15.Info("management server shut down", "error", tun.AsHTTP().Serve(ctx, manageTunnels(ctx, sess)))
+	}
+}
+
+func manageTunnels(ctx context.Context, sess libngrok.Session) http.Handler {
+	tunnels := sync.Map{}
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// GET /create will start a new tunnel with the given provider and
+		// allowed user. The new tunnel ID and url will be returned in the HTTP
+		// response.
+		if r.Method == http.MethodGet && r.URL.Path == "/create" {
+			provider := r.FormValue("provider")
+			allowed := r.FormValue("allow")
+			if provider == "" || allowed == "" {
+				httpResp(w, http.StatusBadRequest, "must supply 'provider' and 'allow' form values\n")
+				return
+			}
+
+			tun, err := sess.StartTunnel(ctx, libngrok.HTTPOptions().
+				WithForwardsTo("dump requests").
```

Comment:
what is this "dump requests" string? The underlying function says it should be an addr? https://github.com/ngrok/ngrok-go/blob/0703564d4eb34c167faa4b3e2e2af9da4de358d2/labeled.go#L23

Reply:
This isn't actually used for anything other than showing up in the ngrok dashboard/api as "this tunnel is forwarding to X"

In typical tunnels, if you run `ngrok http http://my-local-service:8080`, that final string is what gets plumbed up. The agent normalizes the address some (so iirc `ngrok http 80` is `ForwardsTo: http://localhost:80` or similar).